BUG: Respect usecols even with empty data

gfyoung opened the pull request:
closes #12493

 in which the `usecols` argument was not being respected for empty data.  This is because no filtering was applied when the first (and only) chunk was being read.

gfyoung committed:
BUG: Respect usecols even with empty data

Closes gh-12493.

gfyoung commented:
Tests are passing.  Should be good to merge.

jreback commented on pandas/io/tests/test_parsers.py:
```diff
@@ -3461,6 +3462,19 @@ def test_buffer_rd_bytes(self):
+
+        def check_usecols(stringIO):
+            df = read_csv(stringIO, names=names, usecols=usecols)
+            assert_array_equal(df.columns, usecols)
```
NEVER use ``numpy.testing`` imports!!!!! 

ALWAYS use pandas testing functions. These are much more comprehensive and fully test all behavior. It is also confusing to a reader.

 Fully test vs an expected frame and use ``assert_frame_equal``

gfyoung replied:
I think I got it with the first exclamation point. ;) - fixed.

gfyoung committed:
BUG: Better handling of empty data reads with Python engine

In Python, when reading an empty file, it used to throw a StopIteration
error with no error message. This PR helps to differentiate the case
when no columns are inferable, which now leads to an EmptyDataError for
both the C and Python engines.

[ci skip]

gfyoung commented:
If some could this build <a href="https://travis-ci.org/pydata/pandas/builds/113483314">#18023</a> (it's an old build), that would be great.  Thanks!

gfyoung commented:
@jreback : Tests are passing once more.  Should be good to merge AFAICT.

jreback commented on pandas/io/parsers.py:
```diff
@@ -1677,10 +1686,27 @@ def _infer_columns(self):
+                    while self.line_pos <= hr:
+                        line = self._next_line()
+
+                except StopIteration:
```
why are you adding all of this code and making this very complicated?

the idea is to make things simpler. pls see if you can find a general soln.

gfyoung replied:
How complicated is this code?  It's essentially boilerplate and logic from the C engine code.  I'm not writing anything new here.  I don't understand why people would have a hard time understanding this code but yet the C engine code is considered comprehensible.

jreback commented on pandas/parser.pyx:
```diff
@@ -1208,7 +1208,16 @@ cdef class TextReader:
                 return None
 
-class CParserError(Exception):
+
```
what are you doing here? you are making things more complicated. now we have different kinds of errors from different parsers? how is that making things simpler

gfyoung replied:
There are two options.  As I wrote above, the "new" code I added is just a mirror of the C engine code, which throws a `CParserError` when the header indices don't match where the reader stops.  So either one: you change `CParserError` to a more generic sounding `ParserError` or you add `PythonParserError`.

jreback commented on pandas/io/html.py:
```diff
@@ -741,7 +741,7 @@ def _parse(flavor, io, match, header, index_col, skiprows,
                                       parse_dates=parse_dates,
                                       tupleize_cols=tupleize_cols,
                                       thousands=thousands))
-        except StopIteration:  # empty table
```
this is not a good idea at all (same as above)

gfyoung replied:
Telling me it's not a very good idea is not very helpful feedback given that that was what was written before in the first place.  First, why is it not a good idea?  And second, what might you suggest as alternatives to doing a `try-except` block?  Direct checks of whether the table is empty come to mind, but if there are others, it would be good to know

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -1201,3 +1201,4 @@ Bug Fixes
 - Bug when initializing categorical series with a scalar value. (:issue:`12336`)
 - Bug when specifying a UTC ``DatetimeIndex`` by setting ``utc=True`` in ``.to_datetime`` (:issue:`11934`)
 - Bug when increasing the buffer size of CSV reader in ``read_csv`` (:issue:`12494`)
```
this would be for 0.18.1

jreback commented:
@gfyoung you had to change a whole bunch of things to get your code to work (e.g. changing long established behavior in excel/html parsing), the ``StopIteration`` can now also be a ``ValueError``. That is just crazy. My point is to make minimal changes that get the job done, use general methods where possible, and don't duplicate code.

gfyoung commented:
@jreback : Simplified things a little by changing the error from `ValueError` to `StopIteration`.  I still kept some of the error messages from the C engine because I find them to be quite useful.

jreback commented:
at a glance looks ok. still working on 0.18.0, so will get to next week.

gfyoung commented:
:+1: will ping on all of them next week then.

jreback commented on pandas/io/tests/test_parsers.py:
```diff
@@ -2664,6 +2664,35 @@ def test_empty_header_read(count):
+        # throws the correct error, with or without usecols
+        name = self.__class__.__name__
+        errmsg = "No columns to parse from file"
+        klass = (ValueError if 'CParser' in name else StopIteration)
```
why are you doing this? Why can't you set the error message absolutley for each case?

gfyoung replied:
It's the same reason as before.  `CParserError` is engrained into the C engine.  I guess we could have `CParserError` also inherit from `StopIteration` then?

gfyoung replied:
Actually, scrap that idea of mine.  That causes a lot of test failures when I do that.

jreback replied:
no, see my note above. I would like to fix this; only a couple of places in excel/html to fix (and its just changing the exception to catch)

jreback commented on pandas/io/tests/test_parsers.py:
```diff
@@ -3349,6 +3378,12 @@ def test_whitespace_lines(self):
 
+    def test_read_with_bad_header(self):
+        errmsg = "but only \d+ lines in file"
+        with tm.assertRaisesRegexp(StopIteration, errmsg):
```
this is an API change, why are you doing it like this?

jreback replied:
I c this is different between the c and python parser.

gfyoung replied:
Indeed.  Unfortunately, the exceptions that are thrown on the C engine and Python engine sides have historically been different as you referenced below.

jreback replied:
see what u can do to sync these up and raise s CParserError (for the c parser) or a ValueError for the same op in the python parser - use the oarser as the guide

this maybe pretty tricky as u have to track each change

gfyoung replied:
Okay, now I think I understand what you mean by "syncing" them.  Will give that a shot.

gfyoung replied:
On second look, it's not that significant of a change because a `StopIteration` would have been raised before, just without any informative error message.

jreback commented:
don't continue to propogate the differnces between the c and python parser. Let's sync them up. The c-parser is almost certainly right (the cases are of course when a ``ValueError/CParserError`` is throw vs a ``StopIteration``) which is for a completely different reason.

For each thing that is changing, you need to be VERY explict in the whatsnew note, showing an example of where a change is made.

gfyoung commented:
What do you mean by "syncing" them up?  I'm a little stuck here because you said above that we should keep the changes as simple as possible.

gfyoung commented:
Actually, I just realized that once #12551 is landed, aligning `CParserError` with `ValueError` will make a lot more sense since `CParserError` will then inherit from `ValueError`.

gfyoung commented:
@jreback : I've attempted to sync up `CParserError` with `ValueError` on the Python side, but the problem is that it seems `StopIteration` is too ingrained into the code at the moment.  Converting those errors to `ValueError` suddenly causes breakages all over the place in `test_parsers.py` and other files like `test_html.py` or `test_excel.py`.

While I haven't made the commit here, I was able to locally change all the `CParserError`'s to  plain `ValueError`'s with no issues.  However, that is breaking backwards compatibility to a degree.  What do you think is the best way to proceed?

jreback commented on pandas/io/parsers.py:
```diff
@@ -1681,10 +1690,25 @@ def _infer_columns(self):
+
+                except StopIteration:
+                    if self.line_pos < hr:
+                        raise StopIteration(
```
I would just change these to ``ValueError`` (and fix the issue in the html/excel parser, should just mean catching a different exception). This would greatly align the c/python parser errors. (same on line 1709)

jreback commented on pandas/io/parsers.py:
```diff
@@ -1749,10 +1773,18 @@ def _infer_columns(self):
+
+            except StopIteration:
+                if not names:
+                    raise StopIteration("No columns to parse from file")
```
same here

jreback commented on pandas/io/tests/test_parsers.py:
```diff
@@ -2695,6 +2695,44 @@ def test_uneven_lines_with_usecols(self):
+    def test_read_with_bad_header(self):
+        name = self.__class__.__name__
+        errmsg = "but only \d+ lines in file"
+        klass = (ValueError if 'CParser' in name else StopIteration)
```
we DO NOT want things like this

gfyoung commented:
@jreback : It's not as simple as you make it or sound, or as I originally thought.  I wouldn't have made that comment if it was easy to do.  Changing that `StopIteration` alone breaks a good number of tests not just for `html.py` and `excel.py` but for `parsers.py` as well.  I will give it another shot today, but I think this change should be refrained for a separate PR.

gfyoung commented:
Also, any comment on whether this PR _should_ include the conversion of `CParserError` to `ValueError`?

jreback commented:
no, you should not change ANY ``CParserErrors`` to ``ValueError``, it shouldn't be necessary. These are external errors. The issue ATM is that c-parser and python-parser return DIFFERENT errors. The ``CParserError`` (or ``ValueError``) is correct and NOT ``StopIteration``

gfyoung commented:
I am running into an issue with the HTML parser in this test <a href="https://github.com/pydata/pandas/blob/master/pandas/io/tests/test_html.py#L629">here</a>.  It used to throw a `CParserError` but because I'm catching `ValueError` instead of `StopIteration`, the `CParserErrror` is swallowed, causing an empty list of tables to be returned as expected with no exceptions being thrown.  Thoughts?

jreback commented:
@gfyoung I would just change that test, a ``ValueError`` should be thrown.

gfyoung commented:
A `ValueError` will never thrown because `html.py` catches them instead of `StopIteration`.

jreback commented:
no, what I mean is that INSIDE html.py you will need to catch the ``ValueError`` and re-raise as (or let it bubble up). These should all have exactly the same semantics. I think they got somewhat muddled because of the differences between c/python parser.

gfyoung commented:
How can I tell the difference between a deliberately raised `ValueError` and ones that are just caught and can be ignored?  Is checking the error message a valid option?

jreback commented:
Do you *need* to catch them? I don't really recall in that code.

generally we really really try to avoid doing that kind of checking.

gfyoung commented:
Yes.  The code deliberately catches `StopIteration` because it means there is an empty table (there is even a comment next to the `catch` statement to explain why)

jreback commented:
so is the issue that you can't tell the difference between an empty table or one that simply could not be read? aren't these functionaily equivalent from html/excel perspective anyhow?

gfyoung commented:
Yes, when I change it to `ValueError`, I can no longer tell the difference between malformed and empty.  I am not entirely sure at the moment.  From a user perspective, I would think raising on a malformed table makes sense, but an empty table isn't necessarily something that I would "punish" the user for having though.

gfyoung commented:
Yes, when I change it to `ValueError`, I can no longer tell the difference between malformed and empty.  I am not entirely sure at the moment what is best.  From a user perspective, I would think raising on a malformed table makes sense, but an empty table (especially in an HTML file) isn't necessarily something that I would "punish" the user for having though.

jreback commented:
Let's go ahead and change it and just catch/ignore the ``ValueError``. This is ONLY because we are using the python parser, which doesn't make much sense. These should be the same.

gfyoung commented:
Sounds good.  I'll also put in API changes for both `html.py` and `excel.py`.

jorisvandenbossche commented:
I didn't follow the discussion in this PR (so please bear my ignorance), but I just read the whatsnew note, and I have to say that it's not fully clear to me:

> The HTML and Excel parsers no longer differentiate between empty data and parsing errors and will return an empty tables if either situation occurs

Does this exactly mean what is said there? So that if you have for some reason an error parsing a file with data, you silently get back an empty dataframe instead of being notified of this error? 
How is that an improvement?

gfyoung commented:
@jorisvandenbossche : If you read the conversation above, that could give you a better context as to why that decision was made.

@jreback : Travis has given the green light with the changes to `html.py` and `excel.py` so this is ready to land if there is nothing else.

jreback commented on pandas/io/tests/test_html.py:
```diff
@@ -627,11 +626,12 @@ def test_parse_dates_combine(self):
-                                   "too many rows for this multi_index "
-                                   "of columns"):
-            self.read_html(data, header=[0, 1])
+        res = self.read_html(data, header=[0, 1])
```
@gfyoung no we don't want to swallow legit errors
this was not to change the user api

gfyoung replied:
Okay, I think I misunderstood the first time around.  I'll give it another go.

jreback commented on pandas/io/tests/test_parsers.py:
```diff
@@ -2690,6 +2690,42 @@ def test_uneven_lines_with_usecols(self):
+        # first, check to see that the response of
+        # parser when faced with no provided columns
+        # throws the correct error, with or without usecols
+        name = self.__class__.__name__
```
why do you add extraneous things like name?

gfyoung replied:
That was a vestige from when I had the `klass = CParser if...else StopIteration`.  I'll remove it.

jreback commented:
@gfyoung so the way to actually do this properly and not swallow errors
is the make a EmptyParsingError(ValueError) or something like this
and raise it (and catch at the appropriate place)

then ValueErrors will bubble up and you can catch empty partings if needed

gfyoung commented:
@jreback , @jorisvandenbossche : I think I got it this time around, and Travis gives the green light.  If there is nothing else, this is ready to merge.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -76,8 +76,7 @@ API changes
 
-
-
+- In ``read_csv``, the Python parser now raises an ``EmptyDataError`` in response to empty columns or header (:issue:`12506`)
```
not what i meant. You are NOT to change the user experience at all.

gfyoung replied:
The previous user experience used to be `StopIteration`, are you saying that we continue to raise that?

jreback replied:
are you sure? show me an explicit example. This is the point, we are not making ANY API changes unless they are very explicit. I suspect some API is not actually tested. In that case you have to make doubly sure by constructing tests whenever you change things (and then of course add the tests).

My entire point of this issue is to really minimize the user changes if any. At the same time we have a fundamental disconnect between c-engine and python-engine. We are changing from raiing ``StopIteration`` on *some* errors to a ``ValueError``. So this is  small API change that will have to be shown with a complete example. But I don't think we really have coverage of which errors are raised when, so need to sort them out and see what potential changes are. Then we can just whether to accept them now (or defer or whatever).

gfyoung replied:
You can easily find this empty `StopIteration` as follows:
~~~python
>>> from pandas.compat import StringIO
>>> from pandas import read_csv
>>> data = ""
>>> read_csv(StringIO(data), engine='c')
...
ValueError: No columns to parse from file
>>> read_csv(StringIO(data), engine='python') # master
...
StopIteration
~~~
Using my PR:
>>> read_csv(StringIO(data), engine='python') # master
...
pandas.io.parsers.EmptyDataError: No columns to parse from file
~~~

jreback replied:
and is that an actual test that you had to change? I think the answer is no, so pls add that to test coverage. you have to be extraordinarily careful when attempting to change exceptions where we may not necessarily have tests coverage.

gfyoung replied:
No, I added that test.

jreback replied:
I didn't see you specifically catch ``EmptyDataError`` though (IIRC you caught ``ValueError``)

gfyoung replied:
That's because the C engine raises a generic `ValueError` in the analogous place.

jreback commented on pandas/io/excel.py:
```diff
@@ -465,7 +465,7 @@ def _parse_cell(cell_contents, cell_typ):
                     output[asheetname].columns = output[
                         asheetname].columns.set_names(header_names)
-            except StopIteration:
+            except EmptyDataError:
```
you need to raise this at a lower level and catch the appropriate ValueError/EmptyDataError here

gfyoung replied:
I don't quite understand what you mean by that.

jreback replied:
there are 2 cases here, a ``ValueError`` and an ``EmptyDataError``. you should catch the ``EmptyDataError`` and continue, but NOT a plain ``ValueError``

gfyoung replied:
Isn't that what I'm currently doing?

jreback replied:
no

jreback replied:
not HERE

jreback replied:
this is a USER of the parser. You simply changed the error message. That doesn't solve anything. You need to differentiate them.

gfyoung replied:
What do you mean differentiate?  I thought I was supposed to catch an error specific to empty data, which is what the PR does.  It's not "just" a change of error message.

jreback replied:
well now you are ALWAYS raising ``EmptyDataError``, but before it would raise ``StopIteration`` AND ``ValueError``. so you are now gloming those together, that is NOT the purpose. you need to have 2 except blocks. This is similar to the idea of checking the messages on an exception but much cleaner.

jreback replied:
The point is you need to put a halt in the current code and see what errors message that are being caught at the ``StopIteration``, and make sure that we are testing for each one. So when its changed you can be sure that you are catching the correct ones.

gfyoung replied:
No I am not.  I am specifically raising `EmptyDataError` in specific locations where an empty `StopIteration` used to be raised.  I am not gloming them together.  The `try-except` will let generic `ValueError` bubble up still.

gfyoung replied:
And yes, I agree that testing needs to make sure that they are raised, which is why I created those tests in my PR.

jreback replied:
prove it by having tests that raise these separtely (this may not exist or partially), that is the point, want to prove that we are only making very deliberate API changes.

IOW, pretend you are a user reading the release notes, I want to know what changes I can expect. So the whatsnew should show exactly what is changing.

gfyoung replied:
Did you read my comment above?  I said I've already done just that.  I wrote the tests initially to surface these specific errors when I was making the changes.

jreback replied:
they don't differentiate

gfyoung replied:
It doesn't differentiate because the C engine throws a generic `ValueError`.

jreback commented on pandas/io/html.py:
```diff
@@ -741,7 +741,7 @@ def _parse(flavor, io, match, header, index_col, skiprows,
                                       thousands=thousands))
-        except StopIteration:  # empty table
+        except EmptyDataError:  # empty table
             continue
```
same here

jreback commented on pandas/io/parsers.py:
```diff
@@ -37,6 +37,15 @@
 ])
 
 
+class EmptyDataError(ValueError):
```
this is essentially an internal error (yes there are specific cases where the user will see it, but not very many)

gfyoung replied:
But if the user _can_ see it, I think the naming can remain as is, yes?

jreback commented on pandas/io/parsers.py:
```diff
@@ -1719,10 +1737,25 @@ def _infer_columns(self):
+                            % (hr, self.line_pos + 1))
+
+                    # We have an empty file, so check
+                    # if columns are provided. That will
```
a caller can now differntiate between these errors.

gfyoung replied:
Okay, right...so what is the purpose of this comment?

jreback commented on pandas/io/tests/test_parsers.py:
```diff
@@ -2690,6 +2690,41 @@ def test_uneven_lines_with_usecols(self):
         df = self.read_csv(StringIO(csv), usecols=usecols)
         tm.assert_frame_equal(df, expected)
 
```
make another tests that shows ``EmptyDataError``. Furthermore the c-parser needs to conform to this as well.

gfyoung replied:
The tests I added will surface this error, and if the C engine is to raise this as well, I'll need to refactor the location of the `EmptyDataError` first and then I can modify the test to specifically test for that error.

jreback commented on pandas/io/tests/test_parsers.py:
```diff
@@ -2690,6 +2690,41 @@ def test_uneven_lines_with_usecols(self):
+        # throws the correct error, with or without usecols
+        errmsg = "No columns to parse from file"
+
+        with tm.assertRaisesRegexp(ValueError, errmsg):
```
This will be a different exception yes?

gfyoung replied:
Similar to your comment above, the C engine raises a generic `ValueError`, while the Python parser would now raise an `EmptyDataError`.

jreback replied:
so part of the purpose of this issue is to fix all of these different conventions. I want the errors for different engines to be the same (with the exception that ``CParserError`` name needs to remain for ``ValueError`` things).

gfyoung replied:
Okay, that is what I suspected we were heading towards, and I think it will clear up a lot of confusion in the higher up conversations.  FYI that is going to be an API change because we are standardizing across engines.

jreback replied:
@gfyoung and that I am ok with. Or at least once we have a nice example we can decide whether its too big for 0.18.1 or not. But this needs to happen.

So all that said I know I said putting these together is a good idea (the usecols fix and error message changes). 

But might make sense to break these out into 2 PRs. (usecols changes on current master).

Then you can base your changes on that PR (for error message consistency).

gfyoung replied:
Let me try refactoring the `EmptyDataError` first and then we can see.

gfyoung commented:
@jreback : I refactored the `EmptyDataError` into `parser.pyx` so that the C engine can use it, and Travis is happy with that change.  So I guess it really depends now on whether we should split the PR or merge it as is.

gfyoung commented:
@jreback : I refactored the `EmptyDataError` into `parser.pyx` so that the C engine can use it, and Travis is happy with that change.  So I guess it really depends now on whether we should split the PR or merge it as is.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -76,8 +76,7 @@ API changes
 
-
-
+- ``read_csv`` now raises an ``EmptyDataError`` in response to empty columns or header (:issue:`12506`)
```
say this is a sub-class of ``ValueError``, which it previously raised, so just a rename change. I would show a full example, before & after to encompass the 'API' changes that are here. IOW, show the previous usage & errors, and the new. (use code-blocks for everything)

gfyoung replied:
I added examples to the whatsnew.  However, as I've never done it before, not sure if I did it entirely correctly.

jreback commented on pandas/io/tests/test_parsers.py:
```diff
@@ -35,6 +35,7 @@
 from pandas.io.common import DtypeWarning, URLError
 from pandas.io.parsers import (read_csv, read_table, read_fwf,
                                TextFileReader, TextParser)
+from pandas.parser import CParserError, EmptyDataError
```
hmm, I think these errors should be defined in ``pandas/io/common.py`` instead

gfyoung replied:
Sure thing.  Done.

jreback commented on pandas/io/tests/test_parsers.py:
```diff
@@ -2690,6 +2691,41 @@ def test_uneven_lines_with_usecols(self):
+        names = ['Dummy', 'X', 'Dummy_2']
+        usecols = names[1:2]  # ['X']
+
+        def check_usecols(stringIO, expected):
```
I don't think this function adds anything, can you just put in-line below

gfyoung replied:
Two different checks, that's fair.  Done.

jreback commented on pandas/parser.pyx:
```diff
@@ -1239,6 +1239,10 @@ class CParserError(ValueError):
 
 
+class EmptyDataError(ValueError):
+    pass
```
I would move both of these (as well as the ``ParserWarning``) to ``io/common.py`` (you *may* need to import them when you are using to avoid circular imports)

gfyoung replied:
Well if we're moving `CParserError` and `EmptyDataError` from `parser.pyx` to `io/common.py`, any thoughts about `OverflowError`?

jreback replied:
yes that too (though odd why we aren't just using plan ``OverflowError``) which is a standard exception.

gfyoung replied:
Any objections to attempting to removing it?

gfyoung replied:
Any objections to attempting to removing it?  A quick search suggests it may not be too painful.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -76,6 +76,25 @@ API changes
 
 
 
+- ``read_csv`` now raises an ``EmptyDataError``, a subclass of ``ValueError``, in response to empty columns or header (:issue:`12506`)
```
make this  sub-section

``changes in read_csv errors``

gfyoung replied:
Done.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -76,6 +76,25 @@ API changes
+
+.. ipython:: python
+
+   from pandas import read_csv
```
don't need the first one, use ``pd.read_csv``

use ``from StringIO import StringIO``

gfyoung replied:
Python 3 users though?

jreback replied:
you don't actually need the import at all. We are already importing this in the docs (and there is a note already). These docs should be really simple / straightforwar. Look at the v0.18.0 and see how the sub-sections are done.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -76,6 +76,25 @@ API changes
+   from pandas import read_csv
+   from pandas.compat import StringIO
+
+   data = ''
```
you need code-blocks (see other whatsnew for how to do it)

gfyoung replied:
Done.

gfyoung commented:
@jreback : Travis is giving the green light, and I got to get rid of `OverflowError` as a bonus.  Ready to merge if there is nothing else.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -75,6 +75,43 @@ API changes
 
 
 
+- ``pandas.parser.OverflowError`` has been removed and has been replaced with Python's built-in ``OverflowError`` (:issue:`12551`)
```
can you show an example where the error is throw (just put in line here)

gfyoung replied:
Hmmm...not really sure how to illustrate it.  Even the test that's supposed to get the error thrown actually throws the _real_ OverflowError now.

jreback replied:
its fine

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -75,6 +75,43 @@ API changes
+Changes in ``read_csv`` errors
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In :issue:`12506`, in an effort to standardize the ``read_csv`` API for both the C and Python engines, both
```
put the issue at the end like we do for everything else

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -75,6 +75,43 @@ API changes
+
+Changes in ``read_csv`` errors
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
```
you don't have to say in an effort to. just say what the change is.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -75,6 +75,43 @@ API changes
+
+   In [1]: df = pd.read_csv(StringIO(''), engine='c')
+   ...
+   pandas.parser.EmptyDataError: No columns to parse from file
```
I though we moved this exception?

gfyoung replied:
Right, good catch.

jreback commented on pandas/io/common.py:
```diff
@@ -55,10 +55,22 @@ def urlopen(*args, **kwargs):
 
+class CParserError(ValueError):
+    pass
+
```
can you put doc-strings for these (e.g. explain when/what they do)

jreback commented on pandas/io/tests/test_parsers.py:
```diff
@@ -2525,9 +2525,8 @@ def test_int64_overflow(self):
-        self.assertRaises((OverflowError, pandas.parser.OverflowError),
-                          self.read_csv, StringIO(data),
-                          converters={'ID': np.int64})
+        self.assertRaises(OverflowError, self.read_csv,
```
i see (so don't worry about the example above)

jreback commented:
as long as we are changing things, there are 3 bar ``Exception`` in ``parser.pyx``. They can all be ``ValueError``

gfyoung commented:
Sure thing.  I think though one of those generic `Exception` could actually be `CParserError` given the error message.

gfyoung commented:
@jreback : The errors were changed, and Travis is still happy.  Ready to merge if there is nothing else.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -74,6 +74,58 @@ API changes
 - ``CParserError`` is now a ``ValueError`` instead of just an ``Exception`` (:issue:`12551`)
 
 
+- A ``CParserError`` is now raised instead of a generic ``Exception`` in ``read_csv`` when the C engine cannot parse a column (:issue:`12551`)
```
move all of these to the below subsection as bullet points

just out the pr number once

gfyoung replied:
Fair enough.  Done.

jreback commented on pandas/io/tests/test_parsers.py:
```diff
@@ -2690,6 +2689,37 @@ def test_uneven_lines_with_usecols(self):
+            self.read_csv(StringIO(''))
+
+        with tm.assertRaisesRegexp(EmptyDataError, errmsg):
+            self.read_csv(StringIO(''), usecols=usecols)
```
I suspect there are tests that are testing for Exception but now need to look for the specific exception

gfyoung replied:
You are indeed right.  I changed some of those to `ValueError` now.

gfyoung commented:
@jreback : Modified some of the tests that initially tested for `Exception`, and Travis is still happy.  Ready to merge if there is nothing else.

jreback commented:
@jorisvandenbossche @TomAugspurger pls have a look

TomAugspurger commented on pandas/io/tests/test_parsers.py:
```diff
@@ -2690,6 +2689,37 @@ def test_uneven_lines_with_usecols(self):
         df = self.read_csv(StringIO(csv), usecols=usecols)
         tm.assert_frame_equal(df, expected)
 
+    def test_read_empty_with_usecols(self):
```
Thoughts on running this test under both `engine='c'` and `engine='python'? Are we that confident that any inconsistencies have already been worked out before we get here?

jreback replied:
the self.read_csv runs under all engines already

gfyoung commented:
@jreback : Is this good to merge, or are we still waiting to see if there are anymore complaints?

jreback commented:
@jorisvandenbossche review

gfyoung commented:
@jreback : I haven't seen any complaints from @jorisvandenbossche in the past week...so...good to go?

jreback commented:
@gfyoung no I don't think he's had a chance to look.

gfyoung commented:
@jreback : Okay, I'll ping again in about a week then.

gfyoung commented:
@jorisvandenbossche : Any update on this PR?  I think @jreback is waiting for your feedback before he merges this in.

gfyoung commented:
@jorisvandenbossche : Any update on this PR?  I think @jreback is waiting for your feedback before he merges this in.

gfyoung commented:
@jorisvandenbossche : Any update on this PR?  I think @jreback is waiting for your feedback before he merges this in.

jorisvandenbossche commented:
Sorry, slowly getting back on track :-) I will take a look now. 

Just a quick question, my previous comment was about HTML and Excel parser. The whatsnew note now does not mention those anymore. There are no longer changes to to `read_html` and `read_excel` ?

gfyoung commented:
@jorisvandenbossche : No worries. :smile: They are in fact changes to `read_html` and `read_excel`.  If you look at the changed files, both changed the exception they are catching from `StopIteration` to `EmptyDataError`.

jorisvandenbossche commented on pandas/io/tests/test_parsers.py:
```diff
@@ -2796,6 +2795,37 @@ def test_usecols_with_integer_like_header(self):
+        df = self.read_csv(StringIO(''), names=names, usecols=usecols)
+
+    def test_read_with_bad_header(self):
+        name = self.__class__.__name__
```
is this used?

gfyoung replied:
Wait...why did `flake8` not catch that?  No, I don't think so AFAICT.

jorisvandenbossche commented:
@gfyoung But they only catch it under the hood? (I mean, it doesn't bubble up to the user?)

jorisvandenbossche commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -179,6 +179,43 @@ New Behavior:
+
+   In [1]: df = pd.read_csv(StringIO(''), engine='c')
+   ...
+   pandas.io.common.EmptyDataError: No columns to parse from file
```
Is it actually shown with the full name? (just a question, didn't test it, didn't fetch the PR, but I would just show the same as in an actual console)

gfyoung replied:
Which is what I did. :smile: - FYI you can observe this full out name thing if you trigger any current `CParserError`.

jorisvandenbossche replied:
OK, perfect! (I just wondered if it was the case)

gfyoung commented:
@jorisvandenbossche : Not AFAICT (it's an explicit `try`-`except` block for `read_html` and `read_excel`)

jorisvandenbossche commented:
@jreback @gfyoung I didn't look at the code changes in detail, but the whatsnew is in any case very clear and sounds logical! So good to go for me

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -179,6 +179,43 @@ New Behavior:
     # Output is a DataFrame
     df.groupby(pd.TimeGrouper(key='date', freq='M')).apply(lambda x: x[['value']].sum())
 
+Changes in ``read_csv`` errors
```
this sub-section should be in API changes.

add a direction above for referncding this: ``.._whatsnew.......``

jreback replied:
say ``Change in ``read_csv`` exceptions``

gfyoung replied:
Done.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -179,6 +179,43 @@ New Behavior:
+
+- ``CParserError`` is now a ``ValueError`` instead of just an ``Exception`` (:issue:`12551`)
+- A ``CParserError`` is now raised instead of a generic ``Exception`` in ``read_csv`` when the C engine cannot parse a column
+- A ``ValueError`` is now raised instead of a generic ``Exception`` in ``read_csv`` when the C engine encounters a NaN value in an integer column
```
put ``NaN`` in double backticks

gfyoung replied:
Done.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -179,6 +179,43 @@ New Behavior:
+- ``CParserError`` is now a ``ValueError`` instead of just an ``Exception`` (:issue:`12551`)
+- A ``CParserError`` is now raised instead of a generic ``Exception`` in ``read_csv`` when the C engine cannot parse a column
+- A ``ValueError`` is now raised instead of a generic ``Exception`` in ``read_csv`` when the C engine encounters a NaN value in an integer column
+- A ``ValueError`` is now raised instead of a generic ``Exception`` in ``read_csv`` when ``true_values`` is specified, and the C engine encounters an element in a column containing un-encodable bytes
```
is that how you spell ``un-encodable``? (maybe its unencodeable``?)

gfyoung replied:
It just might be.  Fixed.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -179,6 +179,43 @@ New Behavior:
+- A ``CParserError`` is now raised instead of a generic ``Exception`` in ``read_csv`` when the C engine cannot parse a column
+- A ``ValueError`` is now raised instead of a generic ``Exception`` in ``read_csv`` when the C engine encounters a NaN value in an integer column
+- A ``ValueError`` is now raised instead of a generic ``Exception`` in ``read_csv`` when ``true_values`` is specified, and the C engine encounters an element in a column containing un-encodable bytes
+- ``pandas.parser.OverflowError`` has been removed and has been replaced with Python's built-in ``OverflowError``
```
say ``pandas.parser.OverflowError`` exception........``OverflowError`` exception.

gfyoung replied:
Done.

jreback commented:
@gfyoung ok just a couple of minor doc comments. Pls also have a look thru ``io.rst`` and see if any exceptions are mentioned (and if so fix them), as well as doc-strings.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -179,6 +179,45 @@ New Behavior:
+
+In addition to this error change, several others have been made as well:
+
+- ``CParserError`` is now a ``ValueError`` instead of just an ``Exception`` (:issue:`12551`)
```
was this whatsnew just not put in before? (the PR was already merged)

gfyoung replied:
Yes, but I moved it into this section because it's related

jreback commented:
@gfyoung thanks!

gfyoung commented:
@jreback : Sure thing!  FYI, you can also cancel my build <a href="https://travis-ci.org/pydata/pandas/builds/122648340">here</a>.
